gradle-plugin/src/main/java/com/squareup/anvil/plugin/CheckMixedSourceSet.kt

```diff
@@ -22,44 +23,52 @@ open class CheckMixedSourceSet(
   private val compileTask: KotlinCompile
 ) {
 
-  fun disablePreciseJavaTrackingIfNeeded() {
+  fun shouldDisablePreciseJavaTrackingIfNeeded(): Property<Boolean> {
     val sourceFiles = when {
-      project.isAndroidProject -> getSourceFilesAndroidProject()
-      project.isKotlinJvmProject -> getSourceFilesJavaProject()
+      project.isAndroidProject -> project.getSourceFilesAndroidProject(compileTask)
+      project.isKotlinJvmProject -> project.getSourceFilesJavaProject()
       // Consider Kotlin only project in the future.
       else -> throw GradleException("Only Android and JVM modules are supported for now.")
     }
 
     // If there is Java file, then disable precise Java tracking.
-    val hasJavaFile = sourceFiles.any { it.extension == "java" }
+    val hasJavaFile = project.provider {
+      sourceFiles.any { it.extension == "java" }
+    }
 
     // If Kapt is enabled then it usually generates Java code making Kotlin compilation use a
     // mixed source set.
-    val isKaptApplied = project.plugins.hasPlugin(Kapt3GradleSubplugin::class.java)
-
-    if (hasJavaFile || isKaptApplied) {
-      compileTask.usePreciseJavaTracking = false
+    val isKaptApplied = project.provider {
+      project.plugins.hasPlugin(Kapt3GradleSubplugin::class.java)
     }
+
+    return project.objects.property(Boolean::class.java)
+        .also {
+          it.set(hasJavaFile.get() || isKaptApplied.get())
+        }
   }
 
   @OptIn(ExperimentalStdlibApi::class)
-  private fun getSourceFilesAndroidProject(): Sequence<File> {
-    return project.androidVariants()
-        .findVariantForCompileTask(compileTask)
-        .sourceSets
-        .asSequence()
-        .flatMap { it.javaDirectories.asSequence() }
-        .flatMap { it.walk() }
-        .filter { it.isFile }
-  }
+  private fun Project.getSourceFilesAndroidProject(compileTask: KotlinCompile) =
+    this.files( this.provider {
+      this.androidVariants()
+          .findVariantForCompileTask(compileTask)
+          .sourceSets
+          .asSequence()
+          .flatMap { it.javaDirectories.asSequence() }
+          .flatMap { it.walk() }
+          .filter { it.isFile }
+          .toSet()
+    })
 
-  private fun getSourceFilesJavaProject(): Sequence<File> {
-    return project.convention.getPlugin(JavaPluginConvention::class.java)
+  private fun Project.getSourceFilesJavaProject() = this.files( this.provider {
+    this.convention.getPlugin(JavaPluginConvention::class.java)
```

Comment:
I don't understand how this works. You make this lazy by wrapping it a provider. Inside of the lambda you shouldn't use the project reference anymore, because it's called at execution time. But you actually use it by calling `this.convention`. How is this different than before?

Reply:
I just chatted with Gradle about this. The configuration will evaluate the boolean property, and CoCa will serialize it which leads to 2 issues:
1. the IO operations are performed too early still
2. the build wouldn't be aware of any change in what we evaluate (source file presence) after the configuration is cached once. 
The solution is to distinguish more formally the creation of the file collection we work on, and separate better execution, and io operations. I will work on that now.